src/lib.rs

```diff
@@ -11,28 +11,214 @@ use rand::Rng;
+}
+
 type Cell = (usize, usize);
 type CellSet = HashSet<Cell>;
 
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct Grid {
+    grid: CellSet,
+    bounds: Option<(usize, usize)>,
+}
+
+impl Grid {
+    fn new(b: Option<(usize, usize)>) -> Self {
+        match b {
+            None => Grid {
+                grid: CellSet::new(),
+                bounds: b,
+            },
+            _ => Grid {
+                grid: CellSet::with_capacity(b.unwrap().0 * b.unwrap().1),
+                bounds: b,
+            },
+        }
+
+    }
+    fn insert(&mut self, cell: Cell) {
+        if self.bounds != None {
+            if cell.0 <= self.bounds.unwrap().0 && cell.1 <= self.bounds.unwrap().1 {
+                self.grid.insert(cell);
+            }
+        } else {
+            self.grid.insert(cell);
+        }
+    }
+    fn contains(&self, cell: &Cell) -> bool {
+        self.grid.contains(cell)
+    }
+
+    fn x_bound(&self) -> usize {
+        self.bounds.unwrap().0
+    }
+
+    fn y_bound(&self) -> usize {
+        self.bounds.unwrap().1
+    }
+
+    pub fn gen(&mut self) {
+        if self.bounds == None {
+            return;
+        }
+        self.grid.clear();
+        for x in 0..self.bounds.unwrap().0 {
+            for y in 0..self.bounds.unwrap().1 {
+                if rand::thread_rng().gen_weighted_bool(10) {
+                    self.insert((x, y));
+                }
+            }
+        }
+    }
+}
+
+impl From<Vec<String>> for Grid {
+    /// Returns a Grid interpreted from a string representation
+    ///
+    /// # Arguments
+    ///
+    /// * `s` - Representation of the grid. Each element of the vector
+    /// represents a row in the grid. Hash marks # indicate live cells.
+    /// For example ovec!["#  ", "   ", " # "] represents a grid with live
+    /// cells at (0, 0) and (2, 1).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate hemoglobin; fn main(){
+    ///
+    /// let grid = hemoglobin::Grid::from(ovec!["#  ", "   ", " # "]);
+    /// # }
+    /// ```
+    fn from(s: Vec<String>) -> Self {
+        let mut result = Grid::new(None);
+        for (y, row) in s.iter().enumerate() {
+            for (x, c) in row.chars().enumerate() {
+                if c == '#' {
+                    result.insert((x, y));
+                }
+            }
+        }
+        result
+    }
+}
+
+pub struct Rule {
+    dec: BigInt,
+    bin: BitVec,
+}
+
+impl From<BigInt> for Rule {
+    fn from(x: BigInt) -> Self {
+        Rule {
+            dec: x.clone(),
+            bin: {
+                let result_reversed = BitVec::from_bytes(&x.to_bytes_be().1);
+                let mut result = BitVec::from_elem(512, false);
+                for i in 0..result_reversed.len() {
+                    result.set(i, result_reversed[result_reversed.len() - i - 1]);
+                }
+                result
+            },
+        }
+    }
+}
+
+impl From<String> for Rule {
```

Comment:
Might want to make this `From<&str>` instead so it will work for static strings without having to convert to `String` and allocate memory. (Should still be able to use `String`s because they coerce to `&str` with `&`.)

Reply:
Implementing `From<&str>` causes the compiler to force you to use lifetimes, which I do not yet understand. I tried doing that at first, but I failed, which is why I reverted to String and wrote the `ovec!` macro. If you can figure out how to use lifetimes the change will be appreciated.

Reply:
I managed to get it working :)